File: haskell/Problem35.hs
-- Copyright 2015 Mitchell Kember. Subject to the MIT License.
-- Project Euler: Problem 35
-- Circular primes

module Problem35 where

import Common (isPrime, primes, rotate, digits, undigits)

solve :: Int
solve = length . filter works . takeWhile (< 1000000) $ primes
  where
    works = all (isPrime . undigits) . rotations . reverse . digits
    rotations xs = map (flip rotate xs) [1 .. length xs - 1]

File: c++/src/problem_35.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 35
// Circular primes

#include "common.hpp"

namespace problem_35 {

bool is_circular_prime(long n) {
	const long len = common::n_digits(n);
	const long pow = common::powers_of_ten[len - 1];
	for (long i = 0; i < len; ++i) {
		if (!common::is_prime(n)) {
			return false;
		}
		n = n / 10 + (n % 10) * pow;
	}
	return true;
}

long solve() {
	constexpr long max = 1000000;
	long n = 5;
	long count = 3;
	for (;;) {
		n += 2;
		if (n >= max) {
			break;
		}
		if (is_circular_prime(n)) {
			++count;
		}
		n += 4;
		if (n >= max) {
			break;
		}
		if (is_circular_prime(n)) {
			++count;
		}
	}
	return count;
}

} // namespace problem_35
